src/sema/diagnostics.rs

```diff
@@ -141,6 +144,67 @@ impl Diagnostics {
         self.contents.sort();
         self.contents.dedup();
     }
+
+    /// Find overlapping errors and retain the diagnostic with the smallest range
+    pub fn remove_overlapping(&mut self) {
+        if self.contents.is_empty() {
+            return;
+        }
+
+        type Iv = Interval<usize, usize>;
+
+        // in data val is the diagnostic index
+        let data: Vec<Iv> = self
+            .contents
+            .iter()
+            .enumerate()
+            .map(|(index, diag)| Iv {
+                start: diag.loc.start(),
+                stop: diag.loc.end(),
+                val: index,
+            })
+            .collect();
+
+        let laps = Lapper::new(data);
+
+        let mut remove_list = Vec::new();
+
+        for depth in laps.depth() {
+            // in depth val is the number of entries for the depth
+            if depth.val > 1 {
+                // we have multiple diagnostic for range
+                let mut for_range: Vec<_> = laps.find(depth.start, depth.stop).collect();
+
+                assert_eq!(for_range.len(), depth.val);
+
+                for_range.sort_by(|a, b| {
+                    // prefer errors over warnings
+                    let cmp = self.contents[a.val].level.cmp(&self.contents[b.val].level);
+
+                    if cmp != Ordering::Equal {
+                        cmp
+                    } else {
+                        // else prefer shorter range over longer
+                        (a.stop - a.start).cmp(&(b.stop - b.start))
+                    }
+                });
+
+                // we may up with dups in remove_list, but that's not really a problem
+                for_range[1..].iter().for_each(|v| remove_list.push(v.val));
+            }
+        }
+
+        // remove all diagnostics in remove_list
+        let mut contents = Vec::new();
+
+        swap(&mut self.contents, &mut contents);
+
+        for (index, diag) in contents.into_iter().enumerate() {
+            if !remove_list.contains(&index) {
+                self.contents.push(diag);
+            }
+        }
+    }
```

Comment:
Although this solution of removing duplicate diagnostics based on their location is a working solution, I wonder whether this code is really necessary. In codegen, we have explicitly `Isntr::StorageStore` and `Instr::LoadStorage` that can be used to detect storage mutability. In fact, moving the mutability check to codegen was my original idea for #1544, but I found a quick fix for the problem (which led to other unforeseen errors, unfortunately).

I know this would require a lot more work than this PR, so we might solve this later and I would be happy to implement this. Please, just share your thoughts about this.

Reply:
I don't think this will work in codegen, if we want to be compatible with solc. Take this example:

```solidity
contract C {
	int[] arr;

	function foo() internal pure returns (int[] storage y) {
		y = arr;
	}
}
```
solc says:
```
Error: Function declared as pure, but this expression (potentially) reads from the environment or state and thus requires "view".
```
codegen:

```
# function C::C::function::foo public:false selector:8ae62232ad41e1fb nonpayable:true
# params: 
# returns: int256[] storage y
block0: # entry
	ty:int256[] storage %y = uint32 16
	return %y
```
(There is no `Instr::LoadStorage`)

Reply:
After some more fixes, I think the `fn remove_overlapping()` is entirely unnecessary.

Reply:
It is interesting that solc considers that a view function, even if we do not read anything from storage and only assign a constant value to the return variable.

> After some more fixes, I think the `fn remove_overlapping()` is entirely unnecessary.

Great! That makes your code much cleaner!